fix(template functions): Use function Title instead of ToTitle

anmoel opened the pull request:
Use function Title instead of ToTitle to get capitalize strings instead of upper strings

Using `cases.Title` because the `strings.Title` function is deprecated and refs to `cases.Title`

resolves #164

hashicorp-cla commented:
[![CLA assistant check](https://cla.hashicorp.com/pull/badge/not_signed)](https://cla.hashicorp.com/hashicorp/terraform-plugin-docs?pullRequest=165)


Thank you for your submission! We require that all contributors sign our Contributor License Agreement ("CLA") before we can accept the contribution. [Read and sign the agreement](https://cla.hashicorp.com/hashicorp/terraform-plugin-docs?pullRequest=165)

[Learn more about why HashiCorp requires a CLA and what the CLA includes](https://www.hashicorp.com/cla)

<sub>Have you signed the CLA already but the status is still pending? [Recheck it](https://cla.hashicorp.com/check/hashicorp/terraform-plugin-docs?pullRequest=165).</sub>

detro commented:
Hey @anmoel - thanks for taking the time to look into this.

I think you are right (`ToString()` is probably not the right choice here), but your implementation I don't believe it would work.

You need to initialize a _caser_, doing something like:

```golang
package main

import (
	"fmt"
	"strings"

	"golang.org/x/text/cases"
	"golang.org/x/text/language"
)

func main() {
	str := "my Odly cAsed striNg"
	caser := cases.Title(language.Und)

	fmt.Println("ToLower: ", strings.ToLower(str))
	fmt.Println("ToUpper: ", strings.ToUpper(str))
	fmt.Println("ToTitle: ", strings.ToTitle(str))
	fmt.Println("Title: ", strings.Title(str))
	fmt.Println("Caser: ", caser.String(str))
}
```

anmoel commented:
I will add a test in my branch and will check this

anmoel commented:
Hi @detro,

i fixed the code and added tests for some template functions.
is there a reason why you use go cmp instead of testify/assert to compare objects in tests?

detro commented on internal/provider/template_test.go:
```diff
@@ -0,0 +1,43 @@
+import (
+	"testing"
+
+	"github.com/stretchr/testify/assert"
```
It's a long standing position of HC to try to avoid dependency that are not required.

Please use `go-cmp` that is already part of the codebase for these tests.

anmoel replied:
it is done

detro reviewed (changes requested):
Hey, thanks for this.

A few things to address though:

- removal of `testify`
- update of the CHANGELOG to add a new bugfix entry
- version should be `v0.12.1`
- update of the README to point at which function, with which casing, is used for the `title` template function

detro commented on internal/provider/template_test.go:
```diff
@@ -0,0 +1,47 @@
+	"github.com/google/go-cmp/cmp"
+)
+
+func TestRenderStringTemplate(t *testing.T) {
```
thanks so much for those tests. Really appreciated.

detro commented on README.md:
```diff
@@ -168,10 +168,10 @@ using the following data fields and functions:
 | `split`         | Split string into sub-strings, by a given separator (ex. `split .Name "_"`).                      |
-| `title`         | Equivalent to [`strings.ToLower`](https://pkg.go.dev/strings#ToTitle).                            |
+| `title`         | Equivalent to [`cases.Title`](https://pkg.go.dev/golang.org/x/text/cases#Title).                  |
 | `tffile`        | A special case of the `codefile` function, designed for Terraform files (i.e. `.tf`).             |
 | `trimspace`     | Equivalent to [`strings.TrimSpace`](https://pkg.go.dev/strings#TrimSpace).                        |
-| `upper`         | Equivalent to [`strings.ToLower`](https://pkg.go.dev/strings#ToUpper).                            |
+| `upper`         | Equivalent to [`strings.ToUpper`](https://pkg.go.dev/strings#ToUpper).                            |
```
well spotted!

detro commented on CHANGELOG.md:
```diff
@@ -71,13 +77,13 @@ BUG FIXES:
 
-* cmd/tfplugindocs: Use existing Terraform CLI binary if available on PATH, otherwise download latest Terraform CLI binary (https://github.com/hashicorp/terraform-plugin-docs/pull/124).
-* cmd/tfplugindocs: Added `tf-version` flag for specifying Terraform CLI binary version to download, superseding the PATH lookup (https://github.com/hashicorp/terraform-plugin-docs/pull/124).
+* cmd/tfplugindocs: Use existing Terraform CLI binary if available on PATH, otherwise download latest Terraform CLI binary (<https://github.com/hashicorp/terraform-plugin-docs/pull/124>).
+* cmd/tfplugindocs: Added `tf-version` flag for specifying Terraform CLI binary version to download, superseding the PATH lookup (<https://github.com/hashicorp/terraform-plugin-docs/pull/124>).
```
```suggestion
* cmd/tfplugindocs: Use existing Terraform CLI binary if available on PATH, otherwise download latest Terraform CLI binary ([#124](https://github.com/hashicorp/terraform-plugin-docs/pull/124)).
* cmd/tfplugindocs: Added `tf-version` flag for specifying Terraform CLI binary version to download, superseding the PATH lookup ([#124](https://github.com/hashicorp/terraform-plugin-docs/pull/124)).
```

detro commented on CHANGELOG.md:
```diff
@@ -71,13 +77,13 @@ BUG FIXES:
 
-* cmd/tfplugindocs: Swapped `.Type` and `.Name` resource and data source template fields so they correctly align (https://github.com/hashicorp/terraform-plugin-docs/pull/44).
-* schemamd: Switched attribute name rendering from bold text to code blocks so the Terraform Registry treats them as anchor links (https://github.com/hashicorp/terraform-plugin-docs/pull/59).
+* cmd/tfplugindocs: Swapped `.Type` and `.Name` resource and data source template fields so they correctly align (<https://github.com/hashicorp/terraform-plugin-docs/pull/44>).
+* schemamd: Switched attribute name rendering from bold text to code blocks so the Terraform Registry treats them as anchor links (<https://github.com/hashicorp/terraform-plugin-docs/pull/59>).
```
Thanks for spotting those, but the format should be `[#NUM](URL_to_NUM)` (as you correctly used above)

detro commented on CHANGELOG.md:
```diff
@@ -1,3 +1,9 @@
+# 0.12.1 (July 04, 2022)
+
+BUG FIXES:
+
+* template functions: funtion `title` creates capitalize strings instead of upper strings ([#165](https://github.com/hashicorp/terraform-plugin-docs/pull/165)).
+
```
Another PR landed so I think you need to rebase and put this in the `0.13.0` section, but still under `BUG FIXES`.

detro commented:
@anmoel I tried to do the rebasing locally, and it's really just about the CHANGELOG.

Once you rebase, we can merge and then I'll start the release process.

detro commented:
Hi @anmoel - as I'd like to fix the bug, I'm happy to take it upon myself to import your commit locally, do the tweaks and rebasing, and open a new PR that superseeds this one.

If you can't get to it today, I'll try to do this either tomorrow or Monday at the latest.

detro commented:
Superseeded by #168 

Thank you so much for your fix: I have made sure your commit was correctly adjudicated.
